snapshot/pkg/controller/snapshotter/snapshotter.go

```diff
@@ -0,0 +1,943 @@
+	if snapshotDataObj == nil {
+		return fmt.Errorf("Failed to update VolumeSnapshot for snapshot %s: no VolumeSnapshotData.", snapshotName)
+	}
+
+	plugin, err := vs.getPlugin(snapshotName, snapshot)
+	if plugin == nil {
+		return fmt.Errorf("Failed to get volume plugin to wait for snapshot creation %s. %v", snapshotName, err)
+	}
+
+	var newSnapshot *crdv1.VolumeSnapshot = nil
+	var conditions *[]crdv1.VolumeSnapshotCondition = nil
+	var status, newstatus string = "", ""
+	// Wait until the snapshot is successfully created by the plugin or an error occurs that
+	// fails the snapshot creation.
+	for {
+		conditions, _, err = (*plugin).DescribeSnapshot(snapshotDataObj)
+		if err != nil {
+			glog.Warningf("failed to get snapshot %v, err: %v", snapshotName, err)
+			continue
+		}
+		newstatus = vs.getSimplifiedSnapshotStatus(conditions)
+		if newstatus == statusPending {
+			if newstatus != status {
+				status = newstatus
+				glog.V(5).Infof("Snapshot %s creation is not complete yet. Status: [%#v] Retrying...", snapshotName, conditions)
+				// UpdateVolmeSnapshot status
+				newSnapshot, err = vs.UpdateVolumeSnapshot(snapshotName, conditions)
+				if err != nil {
+					glog.Errorf("Error updating volume snapshot %s: %v", snapshotName, err)
+					return fmt.Errorf("Failed to update VolumeSnapshot for snapshot %s: %v", snapshotName, err)
+				}
+			}
+			time.Sleep(createVolumeSnapshotDataInterval)
+			continue
+		} else if newstatus == statusError {
+			return fmt.Errorf("Status for snapshot %s is error.", snapshotName)
+		}
+		glog.Infof("waitForSnapshot: Snapshot %s creation is complete: %#v", snapshotName, conditions)
+
+		newSnapshot, err = vs.UpdateVolumeSnapshot(snapshotName, conditions)
+		if err != nil {
+			glog.Errorf("Error updating volume snapshot %s: %v", snapshotName, err)
+			return fmt.Errorf("Failed to update VolumeSnapshot for snapshot %s: %v", snapshotName, err)
+		}
+
+		ind := len(*conditions) - 1
+		snapDataConditions := []crdv1.VolumeSnapshotDataCondition{
+			{
+				Type:               (crdv1.VolumeSnapshotDataConditionType)((*conditions)[ind].Type),
+				Status:             (*conditions)[ind].Status,
+				Message:            (*conditions)[ind].Message,
+				LastTransitionTime: metav1.Now(),
+			},
+		}
+
+		// Update VolumeSnapshotData status
+		err = vs.UpdateVolumeSnapshotData(snapshotDataName, &snapDataConditions)
+		if err != nil {
+			return fmt.Errorf("Error update snapshotData object %s: %v", snapshotName, err)
+		}
+
+		if newstatus == statusReady {
+			glog.Infof("waitForSnapshot: Snapshot %s created successfully. Adding it to Actual State of World.", snapshotName)
+			vs.actualStateOfWorld.AddSnapshot(newSnapshot)
+			// Break out of the for loop
+			return nil
+		} else if newstatus == statusError {
+			return fmt.Errorf("Failed to create snapshot %s.", snapshotName)
+		}
+	}
+
+	return fmt.Errorf("Snapshot %s is NOT completed successfully.", snapshotName)
```

Comment:
CI govet says unreachable code, we will never break out of the above loop

Reply:
It's intentional. The offending method is being called from a goroutine that actually is meant to either succeed or fail... And wait forever when none of that happens.

Reply:
okay, can safely remove this line then